### ISSUE TITLE
flake: Cloud events unit tests

### ISSUE THREAD

**Issue by abayer** [tektoncd/pipeline#5160]
Both `TestReconcile_CloudEvents` in `pkg/reconciler/pipelinerun/pipelinerun_test.go` with:
```
{Failed      pipelinerun_test.go:6018: config maps: [&ConfigMap{ObjectMeta:{config-defaults  knative-testing    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Data:map[string]string{default-cloud-events-sink: http://synk:8080,},BinaryData:map[string][]byte{},Immutable:nil,}]
          }
          }}
```

...and `TestReconcile_CloudEvents/Run_with_no_condition` in `pkg/reconciler/run/run_test.go` with:
```
{Failed      run_test.go:246: error in test Run with no condition: timed out waiting for 1 more events: []string{"(?s)dev.tekton.event.run.started.v1.*test-run"}
    run_test.go:255: error in test Run with no condition: extra event received: "Context Attributes,\n  specversion: 1.0\n  type: dev.tekton.event.run.started.v1\n  source: /apis///namespaces/foo//test-run\n  subject: test-run\n  id: 3d7cd7c6-777f-466d-9c0e-99bf34c99a3e\n  datacontenttype: application/json\nData,\n  {\n    \"run\": {\n      \"metadata\": {\n        \"name\": \"test-run\",\n        \"namespace\": \"foo\",\n        \"resourceVersion\": \"00001\",\n        \"creationTimestamp\": null\n      },\n      \"spec\": {\n        \"serviceAccountName\": \"\"\n      },\n      \"status\": {\n        \"extraFields\": null\n      }\n    }\n  }\n"
INFO	TestReconcile_CloudEvents/Run_with_no_condition	cloudevent/cloud_event_controller.go:157	cloudevent Context Attributes,
  specversion: 1.0
  type: dev.tekton.event.run.started.v1
  source: /apis///namespaces/foo//test-run
  subject: test-run
  id: 318d26ad-b6a3-4944-9ca4-47918e2b9028
  datacontenttype: application/json
Data,
  {
    "run": {
      "metadata": {
        "name": "test-run",
        "namespace": "foo",
        "resourceVersion": "00001",
        "creationTimestamp": null
      },
      "spec": {
        "serviceAccountName": ""
      },
      "status": {
        "extraFields": null
      }
    }
  }
 already sent}
```

...have been failing sporadically in CI lately.

/kind flake

**Comment by lbernick** [comment]
possible duplicate of #2992

**Comment by lbernick** [comment]
Marking as urgent since I am seeing these unit tests constantly failing and having unit tests flake is not great!

**Comment by abayer** [comment]
👍 - I've tried to reproduce the various event test failures (there's more than just the cloud events ones, though those are the most frequent) locally, and sometimes can, but not nearly frequently enough to verify the only potential fix that's come to mind so far (bumping the timeouts in `test/events/events.go` from 10ms to 20ms)...though I also tend to think that is not likely to fix the flakes anyway.

**Comment by abayer** [comment]
Well, this may be helpful specifically for the cloud events flakiness - I got a data race:

```
WARNING: DATA RACE
Read at 0x00c0001038a3 by goroutine 8:
  testing.(*common).logDepth()
      /usr/local/go/src/testing/testing.go:774 +0x4e4
  testing.(*common).log()
      /usr/local/go/src/testing/testing.go:761 +0x6e
  testing.(*common).Logf()
      /usr/local/go/src/testing/testing.go:807 +0x14
  testing.(*T).Logf()
      <autogenerated>:1 +0x75
  go.uber.org/zap/zaptest.testingWriter.Write()
      /home/prow/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest/logger.go:130 +0x12c
  go.uber.org/zap/zaptest.(*testingWriter).Write()
      <autogenerated>:1 +0x88
  go.uber.org/zap/zapcore.(*ioCore).Write()
      /home/prow/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/core.go:90 +0x196
  go.uber.org/zap/zapcore.(*CheckedEntry).Write()
      /home/prow/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/entry.go:220 +0x2ba
  go.uber.org/zap.(*SugaredLogger).log()
      /home/prow/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:227 +0x13a
  go.uber.org/zap.(*SugaredLogger).Infof()
      /home/prow/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:138 +0x645
  github.com/tektoncd/pipeline/pkg/reconciler/events/cloudevent.SendCloudEventWithRetries.func1()
      /home/prow/go/src/github.com/tektoncd/pipeline/pkg/reconciler/events/cloudevent/cloud_event_controller.go:157 +0x5e4

Previous write at 0x00c0001038a3 by goroutine 28:
  testing.tRunner.func1()
      /usr/local/go/src/testing/testing.go:1246 +0x584
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1265 +0x268
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 8 (running) created at:
  github.com/tektoncd/pipeline/pkg/reconciler/events/cloudevent.SendCloudEventWithRetries()
      /home/prow/go/src/github.com/tektoncd/pipeline/pkg/reconciler/events/cloudevent/cloud_event_controller.go:147 +0x3b0
  github.com/tektoncd/pipeline/pkg/reconciler/events.EmitCloudEvents()
      /home/prow/go/src/github.com/tektoncd/pipeline/pkg/reconciler/events/event.go:82 +0x190
  github.com/tektoncd/pipeline/pkg/reconciler/run.(*Reconciler).ReconcileKind()
      /home/prow/go/src/github.com/tektoncd/pipeline/pkg/reconciler/run/run.go:69 +0x5f0
  github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run.Interface.ReconcileKind-fm()
      /home/prow/go/src/github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run/reconciler.go:53 +0x75
  github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run.(*reconcilerImpl).Reconcile()
      /home/prow/go/src/github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run/reconciler.go:239 +0x82c
  github.com/tektoncd/pipeline/pkg/reconciler/run.TestReconcile_CloudEvents.func1()
      /home/prow/go/src/github.com/tektoncd/pipeline/pkg/reconciler/run/run_test.go:250 +0xc21
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 28 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x9d1
  main.main()
      _testmain.go:45 +0x22b
```

**Comment by abayer** [comment]
Comparing a failed run of a test with a successful one:

Failed:
```
    logger.go:130: 2022-08-11T16:04:08.089Z     INFO    TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-11T16:04:08.090Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:67   Emitting cloudevent for test-run, condition:
 &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T16:04:08.092Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    logger.go:130: 2022-08-11T16:04:08.170Z     INFO    TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-11T16:04:08.170Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:67   Emitting cloudevent for test-run, condition: &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T16:04:08.172Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    run_test.go:255: error in test Run with finished true condition: extra event received: "Context Attributes,\n  specversion: 1.0\n  type: dev.tekton.event.run.successful.v1\n  source: /apis///namespaces/foo//test-run\n  subject: test-run\n  id: 697d6d54-5e82-4417-b1ff-27461844d2a0\n  datacontenttype: application/json\nData,\n  {\n    \"run\": {\n      \"metadata\": {\n        \"name\": \"test-run\",\n        \"namespace\": \"foo\",\n        \"resourceVersion\": \"00001\",\n        \"creationTimestamp\": null\n      },\n      \"spec\": {\n        \"serviceAccountName\": \"\"\n      },\n      \"status\": {\n        \"conditions\": [\n          {\n            \"type\": \"Succeeded\",\n            \"status\": \"True\",\n            \"lastTransitionTime\": null,\n            \"reason\": \"Succeeded\"\n          }\n        ],\n        \"extraFields\": null\n      }\n    }\n  }\n"
```

Succeeded:
```
    logger.go:130: 2022-08-11T12:09:11.940-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:11.940-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T12:09:11.940-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    logger.go:130: 2022-08-11T12:09:11.952-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:11.952-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T12:09:11.952-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    logger.go:130: 2022-08-11T12:09:11.952-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_true_condition	cloudevent/cloud_event_controller.go:157	cloudevent Context Attributes,
          specversion: 1.0
          type: dev.tekton.event.run.successful.v1
          source: /apis///namespaces/foo//test-run
          subject: test-run
          id: a1ada847-9b77-4c9c-bdf5-2b89c8cc9798
          datacontenttype: application/json
        Data,
          {
            "run": {
              "metadata": {
                "name": "test-run",
                "namespace": "foo",
                "resourceVersion": "00001",
                "creationTimestamp": null
              },
              "spec": {
                "serviceAccountName": ""
              },
              "status": {
                "conditions": [
                  {
                    "type": "Succeeded",
                    "status": "True",
                    "lastTransitionTime": null,
                    "reason": "Succeeded"
                  }
                ],
                "extraFields": null
              }
            }
          }
         already sent
=== RUN   TestReconcile_CloudEvents/Run_with_finished_false_condition
    logger.go:130: 2022-08-11T12:09:11.962-0400	INFO	cache/cacheclient.go:42	CACHE CLIENT &{lru:0xc0005f9320 lock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0}}
    logger.go:130: 2022-08-11T12:09:11.965-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-artifact-bucket" config was added or updated: &config.ArtifactBucket{Location:"", ServiceAccountSecretName:"", ServiceAccountSecretKey:"", ServiceAccountFieldName:"GOOGLE_APPLICATION_CREDENTIALS"}
    logger.go:130: 2022-08-11T12:09:11.965-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-artifact-pvc" config was added or updated: &config.ArtifactPVC{Size:"5Gi", StorageClassName:""}
    logger.go:130: 2022-08-11T12:09:11.965-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-defaults" config was added or updated: &config.Defaults{DefaultTimeoutMinutes:60, DefaultServiceAccount:"default", DefaultManagedByLabelValue:"tekton-pipelines", DefaultPodTemplate:(*pod.Template)(nil), DefaultAAPodTemplate:(*pod.AffinityAssistantTemplate)(nil), DefaultCloudEventsSink:"http://synk:8080", DefaultTaskRunWorkspaceBinding:"", DefaultMaxMatrixCombinationsCount:256}
    logger.go:130: 2022-08-11T12:09:11.966-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-observability" config was added or updated: &config.Metrics{TaskrunLevel:"task", PipelinerunLevel:"pipeline", DurationTaskrunType:"histogram", DurationPipelinerunType:"histogram"}
    logger.go:130: 2022-08-11T12:09:11.966-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "feature-flags" config was added or updated: &config.FeatureFlags{DisableAffinityAssistant:false, DisableCredsInit:false, RunningInEnvWithInjectedSidecars:true, RequireGitSSHSecretKnownHosts:false, EnableTektonOCIBundles:false, EnableCustomTasks:false, ScopeWhenExpressionsToTask:false, EnableAPIFields:"stable", SendCloudEventsForRuns:true, AwaitSidecarReadiness:true, EmbeddedStatus:"full"}
    logger.go:130: 2022-08-11T12:09:12.065-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:12.065-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded False  {0001-01-01 00:00:00 +0000 UTC} Cancelled }
    logger.go:130: 2022-08-11T12:09:12.065-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.failed.v1"
    logger.go:130: 2022-08-11T12:09:12.076-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:12.076-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded False  {0001-01-01 00:00:00 +0000 UTC} Cancelled }
    logger.go:130: 2022-08-11T12:09:12.076-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.failed.v1"
    logger.go:130: 2022-08-11T12:09:12.077-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_false_condition	cloudevent/cloud_event_controller.go:157	cloudevent Context Attributes,
          specversion: 1.0
          type: dev.tekton.event.run.failed.v1
          source: /apis///namespaces/foo//test-run
          subject: test-run
          id: 12f13e3d-f297-41d0-a15a-1014a7768f1a
          datacontenttype: application/json
        Data,
          {
            "run": {
              "metadata": {
                "name": "test-run",
                "namespace": "foo",
                "resourceVersion": "00001",
                "creationTimestamp": null
              },
              "spec": {
                "serviceAccountName": ""
              },
              "status": {
                "conditions": [
                  {
                    "type": "Succeeded",
                    "status": "False",
                    "lastTransitionTime": null,
                    "reason": "Cancelled"
                  }
                ],
                "extraFields": null
              }
            }
          }
         already sent
```

**Comment by abayer** [comment]
I think I may have it - it looks like if something goes wrong with sending the event, it'll do an exponential backoff retry to resend it. I have a feeling that once in a while, probably due to parallel load, at least one attempt to send an event fails initially and is retried, but the reconciliation ends up finishing before the event sending retries have actually succeeded. I'm experimenting with adding a 50ms delay in `CheckEventsOrdered` and `CheckEventsUnordered` to see if that makes a difference - early signs are that it does.

**Comment by abayer** [comment]
Ok! I finally threw enough debug logging into the code to get something really useful:

```
    logger.go:130: 2022-08-12T14:08:00.771Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-12T14:08:00.772Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition run/run.go:67   Emitting cloudevent for test-run, condition: %!s(*apis.Condi
tion=<nil>)
    logger.go:130: 2022-08-12T14:08:01.299Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "
dev.tekton.event.run.started.v1"
CHECKING CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
EXISTING KEYS: []
    logger.go:130: 2022-08-12T14:08:01.381Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-12T14:08:01.382Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition run/run.go:67   Emitting cloudevent for test-run, condition: %!s(*apis.Condi
tion=<nil>)
    logger.go:130: 2022-08-12T14:08:01.384Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "
dev.tekton.event.run.started.v1"
CHECKING CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
EXISTING KEYS: []
    logger.go:130: 2022-08-12T14:08:01.378Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:169        past sending event dev.tekto
n.event.run.started.v1
    logger.go:130: 2022-08-12T14:08:01.386Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:172        Caching sent cloudevent dev.
tekton.event.run.started.v1
ABOUT TO CACHE ATTEMPT
ABOUT TO EVENTKEY
    logger.go:130: 2022-08-12T14:08:01.386Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:169        past sending event dev.tekto
n.event.run.started.v1
GOT NEW CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
    logger.go:130: 2022-08-12T14:08:01.387Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:172        Caching sent cloudevent dev.
tekton.event.run.started.v1
ABOUT TO CACHE ATTEMPT
ABOUT TO EVENTKEY
NEW KEYS: [dev.tekton.event.run.started.v1/run/foo/test-run]
GOT NEW CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
NEW KEYS: [dev.tekton.event.run.started.v1/run/foo/test-run]
    run_test.go:255: error in test Run with no condition: extra event received: "Context Attributes,\n  specversion: 1.0\n  type: dev.tekton.event.run.started.v1\n  source: /apis//
/namespaces/foo//test-run\n  subject: test-run\n  id: dfdea5ad-d200-40b7-9758-e528392f2ab1\n  datacontenttype: application/json\nData,\n  {\n    \"run\": {\n      \"metadata\": {\n
        \"name\": \"test-run\",\n        \"namespace\": \"foo\",\n        \"resourceVersion\": \"00001\",\n        \"creationTimestamp\": null\n      },\n      \"spec\": {\n
 \"serviceAccountName\": \"\"\n      },\n      \"status\": {\n        \"extraFields\": null\n      }\n    }\n  }\n"
```

What this means is that there are two calls to `SendCloudEventWithRetries` going on at the same time - you can tell, because `ABOUT TO CACHE ATTEMPT` should be showing up after `CHECKING CACHE KEY`, not interwoven as they are here. So we end up sending the event twice because it hasn't yet been added to the cache when the second call gets underway!

**Comment by abayer** [comment]
Basically, `SendCloudEventWithRetries` is not thread-safe for sure, at least for `Run` events.

**Comment by afrittoli** [comment]
I guess there are two reconcile operations running in parallel. The cache itself is supposed to be thread safe, so I guess something in the `SendCloudEventWithRetries` breaks the thread safeness?

**Comment by abayer** [comment]
Yeah - the problem is the operations around cache operations not being thread-safe.

**Comment by abayer** [comment]
I'm testing out a quick and dirty hack of putting a mutex in the `Run` reconciler and locking on it for sending cloud events (since we only do caching for `Run`s). That doesn't quite feel like the "right" solution overall, but it should let me verify that some additional locking around sending cloud events does the trick.

**Comment by abayer** [comment]
Ok, that's looking really good - over 30 minutes since I started looping test runs without a failure, which is longer than I've seen before. I'm going to let it run another hour or so to be sure, then I'll open a PR.

**Comment by abayer** [comment]
I've added that change to https://github.com/tektoncd/pipeline/pull/5313/commits/1cdc4fa81100a8c3525069aa5d0caf9bd27f8a36 - mainly to see if anyone has a better idea of where to put the relevant mutex. =)

**Comment by afrittoli** [comment]
https://github.com/tektoncd/pipeline/pull/5313#issuecomment-1213230447

**Comment by abayer** [comment]
More details on what's going on: https://github.com/tektoncd/pipeline/pull/5313#issuecomment-1213251895

My mutex in the `Run` reconciler approach is definitely working (~40 runs in my test setup without a failure, when previously I'd always see one by at least 5-6 runs in), but it's definitely not ideal. Basically, all of `cloudevent.SendCloudEventWithRetries` needs to be locked behind a mutex, as best as I can tell.

**Comment by abayer** [comment]
fwiw, my approach doesn't eliminate the possible race condition around logging I've got in https://github.com/tektoncd/pipeline/issues/5160#issuecomment-1212193986. That race is unrelated to the overwhelming majority of event-related flakes - I just saw it show up for the first time in my looping of `go test -race ...` runs, after ~70 runs - but it is real.
